src/pgen/crooked_pipe.hpp

```diff
@@ -0,0 +1,162 @@
+  const bool do_radiation = artemis_pkg->Param<bool>("do_radiation");
+  const bool do_imc = artemis_pkg->Param<bool>("do_imc");
+  const bool do_moment = artemis_pkg->Param<bool>("do_moment");
+  PARTHENON_REQUIRE(do_gas, "Crooked pipe problem requires gas!");
+  PARTHENON_REQUIRE(!(do_dust), "Crooked pipe problem does not permit dust!");
+  //PARTHENON_REQUIRE(!(do_gas), "Crooked pipe problem does not permit gas!");
+  auto gas_pkg = pmb->packages.Get("gas");
+  const auto eos = gas_pkg->Param<EOS>("eos_d");
+  Real ar = Null<Real>();
+  if (do_moment) {
+    auto rad_pkg = pmb->packages.Get("moments");
+    ar = rad_pkg->Param<Real>("arad");
+  }
+
+  // Initial conditions
+  const Real rho_thin = pin->GetOrAddReal("problem", "rho_thin", 1.0);
+  const Real rho_thick = pin->GetOrAddReal("problem", "rho_thick", 1000.0);
+  const Real t_source = pin->GetOrAddReal("problem", "t_source", 0.3); // put in K?
+  const Real t_init = pin->GetOrAddReal("problem", "t_init", 0.01); // put in K?
+
+  // Allocate sparse
+  auto &md = pmb->meshblock_data.Get();
+  for (auto &var : md->GetVariableVector()) {
+    if (!var->IsAllocated()) pmb->AllocateSparse(var->label());
+  }
+
+  // packing and capture variables for kernel
+  static auto desc =
+      MakePackDescriptor<gas::prim::density, gas::prim::velocity, gas::prim::sie,
+                         rad::prim::energy, rad::prim::flux>(
+          (pmb->resolved_packages).get());
+  auto v = desc.GetPack(md.get());
+  IndexRange ib = pmb->cellbounds.GetBoundsI(IndexDomain::entire);
+  IndexRange jb = pmb->cellbounds.GetBoundsJ(IndexDomain::entire);
+  IndexRange kb = pmb->cellbounds.GetBoundsK(IndexDomain::entire);
+
+  std::vector<std::array<double, 4>> thick_regions = {{3.0,4.0, -1.0,1.0},
+                                                      {-2.0,2.5, -2.0, 0.5}, // extended xl to -2 for thick region above source
+                                                      {-2.0, 2.5, 0.5, 2.0}, // extended xl to -2 for thick region above source
+                                                      {4.5, 7.0, -2.0, 0.5},
+                                                      {4.5, 7.0, 0.5, 2.0},
+                                                      {2.5, 4.5, 1.5, 2.5}};
+
+  std::vector<std::array<double, 4>> thin_source_regions= {{-2.0,0.0,-1.0,1.0}};
+
+  const auto &cpars =
+      pmb->packages.Get("artemis")->template Param<geometry::CoordParams>("coord_params");
+  auto &pco = pmb->coords;
+
+  static auto desc_g = MakePackDescriptor<geom::vol, geom::x1v, geom::x2v, geom::x3v>(
+      (pmb->resolved_packages).get());
+  auto vg = desc_g.GetPack(md.get());
+
+  // Set state vector to initialize:
+  // * source radiation and material temperatureradiation field via t_source
+  // * density of thin regions using rho_thin
+  // * density of thick regions using rho_thick
+  // * initial material and radiation temperature of all cell using t_init
+  if (do_imc) {
```

Comment:
Suppose for moments, this if-check could be moved to be just around the `jaybenne::InitializeRadiation` invocation below.